{% extends "unitbase.html" %}
{% load pytags %}
{# title = SAM #}
{# subtitle = sequence alignment maps #}
{# name = Representing alignments: Sequence Alignment Maps (SAM) #}

{% block body %}

{% md sam %}

A SAM file contains so called information on alignments.

> **Reference:** The [SAM format specification][samspec]. Keep it under your pillow.

[samspec]: https://samtools.github.io/hts-specs/

- - -

### Creating SAM files

Get the reference genome of the 1976 Mayinga Ebola outbreak

    # Set up the various shortcuts
    REF=~/refs/ebola/1976.fa
    R1=SRR1972739_1.fastq
    R2=SRR1972739_2.fastq

    # Obtain and prepare the genome.
    mkdir -p ~/refs/ebola
    efetch -db=nuccore -format=fasta -id=AF086833 > $REF
    bwa index $REF

    # Obtain the data
    fastq-dump -X 1000 --split-files SRR1972739

    # Align the reads.
    bwa mem $REF $R1 $R2 > bwa.sam

The main program used to process SAM files is called samtools:

    samtools

Just as `bwa` (and as a matter of fact both were developed by the same author, Heng Li) samtools
takes commands like `view`:

    samtools view bwa.sam | head

This generates 10 SAM records. Scroll to the right to see all columns.

    SRR1972739.1  83    gi|10141003|gb|AF086833.2|  15684  60  69M32S  =  15600  -153  TTTAGATTTAACAAGATACCGAGAAAATGAATTGATTTATGACAATAATCCTCTAAAAGGAGGACTCAAATGAGATATTGCAATTGAGTCCTCCTTTTAGA	DDDDDEEEEEDEEFFFEDHHHHIJJJJJJJJJJJJJJJJJIGIJJJJJJJJJJJJJJJJJJJIGIGJJJJJJJJJJJJJJIJJJJJJIFHHHHFFFFFCCC	NM:i:2	MD:Z:27C16G24	AS:i:59	XS:i:0	SA:Z:gi|10141003|gb|AF086833.2|,15735,+,33M68S,60,0;
    SRR1972739.1  2115  gi|10141003|gb|AF086833.2|  15735  60  33M68H  =  15600  -136  TCTAAAAGGAGGACTCAATTGCAATATCTCATT	CCCFFFFFHHHHFIJJJJJJIJJJJJJJJJJJJ	NM:i:0	MD:Z:33	AS:i:33	XS:i:0	SA:Z:gi|10141003|gb|AF086833.2|,15684,-,69M32S,60,2;
    SRR1972739.1  163   gi|10141003|gb|AF086833.2|  15600  60  101M    =  15684   153  GTATAATCGTGCTCACCTTCATCTAACTAAGTGTTGCACCCGGGAGGTACCAGCTCAGTACTTAACATACACATCTACATTGGATTTAGATTTAACAAGAT	@BBFFFFFHHHHHJJJJJJJJJJJJJIIIJJGHIIJJJJJJJJIJJJGIJGIJJIJJJIJHHHHHHGFFFFFEEEEEEEDFDDDDDDDDDDEDDEDDDDDD	NM:i:3	MD:Z:0A44A14T40	AS:i:90	XS:i:0
    SRR1972739.2  115   gi|10141003|gb|AF086833.2|  4919   60  101M    =  4863   -104  CAGGCTCCTGCGAATTGGAAACCAGGCTTTCCTCCAGGAGTTCGTTCTTCCACCAGTCCAACTACCCCAGTATTTCACCTTTGATTTGACAGCACTCAAAC	DDDDDDDDDDDDDDDEDDDDCDDEEEEEFFFHHHHHIIIJJJJJJJJJIJHGJJJJIJJIIIHFJJJJJJJJJIHGFJJJJJJJJJJIHHHGHFFFFFBBB	NM:i:1	MD:Z:51G49	AS:i:96	XS:i:0

### SAM file structure

Cut out one column at the time and read what the [SAM format specification][sam-spec] states about
it.

    # First column is called QNAME (Query Name)
    samtools view bwa.sam  | cut -f 1 | head
    # SRR1972739.1
    # SRR1972739.1
    # SRR1972739.1
    # SRR1972739.2
    # SRR1972739.2

    # Second column is called FLAG
    samtools view bwa.sam | cut -f 2 | head
    # 83
    # 2115
    # 163
    # 115
    #179

    # And so on. Investigate the 11+ columns

The book version of the handbook explains in more detail what each of these flags mean.

### Sam Flags

The samtools `flags` allow us to select alignments that have certain properties.
As of samtools 1.2 there is a `flags` command to covert the integer codes to human readable output:

    # List all flags.
    samtools flags

    # It can combine the flags into a single number.
    samtools flags PAIRED,PROPER_PAIR,REVERSE,MREVERSE
    # 0x33    51    PAIRED,PROPER_PAIR,REVERSE,MREVERSE

    # It works backwards from integer number to words as well.
    samtools flags 51
    # 0x33	51	PAIRED,PROPER_PAIR,REVERSE,MREVERSE

When we filter we subselect from the alignment based on the condition.

    # How many alignments do we have
    samtools view bwa.sam | wc -l
    # 2073

    # Counting alignments is such a common operations that there
    # is a parameter -c to do it
    samtools view -c bwa.sam
    # 2073

### Converting to BAM format

The BAM conversion can be performed via:

    samtools view -b bwa.sam > bwa.bam

Samtools will works identically on BAM file as well, it will detect the
type of the file automatically:

    samtools view bwa.bam | head

Most analyses require that the BAM file is sorted by coordinates and is indexed
as well:

    samtools sort bwa.bam > bwa.sorted.bam
    samtools index bwa.sorted.bam

It is possible to create a one step conversion and sort (make sure you have samtools 1.3 or later):

    alias bam='samtools view -b - | samtools sort'

With this you can chain the `bam` command to the outputs that produce SAM formats
and bypass having the extra files:

    bwa mem $REF $R1 $R2 | bam > bwa.bam
    samtools index bwa.bam

### Filtering BAM files

To select all reads that align and are mapped to the reverse strand we do a:

	# Oppose unmapped (4) and select for reverse strand (16).
	samtools view -c -F 4 -f 16  bwa.bam
	# 651

Now to select the reads that map onto the forward strand we do:

	# Oppose unmapped (4) and oppose the reverse strand (16).
	samtools view -c -F 4 -F 16  bwa.bam
	# 898

	# The two flags can be combined into a single one:
	samtools view -c -F 20  bwa.bam
	# 898

	samtools flags 20
	# 0x14	20	UNMAP,REVERSE

### Visualize BAM files

Sorted and indexed BAM files may be visualized in tools with graphical user interfaces
for example [IGV](https://www.broadinstitute.org/igv/)  or [IGB](http://bioviz.org/igb/index.html).
These tools may already have the genome of interest. If not the users need to create a custom
genome file to visualize against. Please consult the documentation on how to do that.

{% img "SAM-screenshot35.png" css="img-responsive"  %}

{%  endmd %}


{% endblock %}
